linera-views/src/views/historical_hash_wrapper.rs

```diff
@@ -0,0 +1,528 @@
+        let context = context.clone_with_base_key(base_key);
+        v.extend(W::pre_load(&context)?);
+        Ok(v)
+    }
+
+    fn post_load(context: Self::Context, values: &[Option<Vec<u8>>]) -> Result<Self, ViewError> {
+        let hash = from_bytes_option(values.first().ok_or(ViewError::PostLoadValuesError)?)?;
+        let base_key = context.base_key().base_tag(KeyTag::Inner as u8);
+        let context = context.clone_with_base_key(base_key);
+        let inner = W::post_load(
+            context,
+            values.get(1..).ok_or(ViewError::PostLoadValuesError)?,
+        )?;
+        Ok(Self {
+            _phantom: PhantomData,
+            stored_hash: hash,
+            hash,
+            inner,
+        })
+    }
+
+    async fn load(context: Self::Context) -> Result<Self, ViewError> {
+        let keys = Self::pre_load(&context)?;
+        let values = context.store().read_multi_values_bytes(keys).await?;
+        Self::post_load(context, &values)
+    }
+
+    fn rollback(&mut self) {
+        self.inner.rollback();
+        self.hash = self.stored_hash;
+    }
+
+    async fn has_pending_changes(&self) -> bool {
+        self.inner.has_pending_changes().await
+    }
+
+    fn flush(&mut self, batch: &mut Batch) -> Result<bool, ViewError> {
+        let mut inner_batch = Batch::new();
+        self.inner.flush(&mut inner_batch)?;
+        let hash = self.make_hash(&inner_batch)?;
+        batch.operations.extend(inner_batch.operations);
+        if self.stored_hash != Some(hash) {
+            let mut key = self.inner.context().base_key().bytes.clone();
+            let tag = key.last_mut().unwrap();
+            *tag = KeyTag::Hash as u8;
+            batch.put_key_value(key, &hash)?;
+            self.stored_hash = Some(hash);
+        }
+        // Remember the hash.
+        self.hash = Some(hash);
+        // Never delete the stored hash, even if the inner view was cleared.
+        Ok(false)
+    }
+
+    fn clear(&mut self) {
+        self.inner.clear();
+        self.hash = None;
+    }
+}
+
+impl<W> ClonableView for HistoricallyHashableView<W::Context, W>
+where
+    W: ClonableView,
+{
+    fn clone_unchecked(&mut self) -> Result<Self, ViewError> {
+        Ok(HistoricallyHashableView {
+            _phantom: PhantomData,
+            stored_hash: self.stored_hash,
+            hash: self.hash,
+            inner: self.inner.clone_unchecked()?,
+        })
+    }
+}
+
+impl<W: ClonableView> HistoricallyHashableView<W::Context, W> {
+    /// Obtain a hash of the history of the changes in the view.
+    pub async fn historical_hash(&mut self) -> Result<HasherOutput, ViewError> {
+        if let Some(hash) = self.hash {
+            return Ok(hash);
+        }
+        let mut batch = Batch::new();
+        if self.inner.has_pending_changes().await {
+            let mut inner = self.inner.clone_unchecked()?;
+            inner.flush(&mut batch)?;
```

Comment:
Doesn't this modify `inner`?
If we call `historical_hash` and then `flush`, doesn't that store a different state than just calling `flush`? Maybe this could be checked in `test_historically_hashable_view_hash_changes_with_modifications`.

Reply:
That's why `inner` is a deep-copy of `self.inner`. Am I misunderstanding the question?